_posts/2021-03-19-forecasting-time-series-with-torch_4/forecasting_4.Rmd

```diff
@@ -0,0 +1,570 @@
+        hidden_size = hidden_size,
+        num_layers = num_layers,
+        dropout = dropout,
+        batch_first = TRUE
+      )
+    } else {
+      nn_lstm(
+        input_size = input_size,
+        hidden_size = hidden_size,
+        num_layers = num_layers,
+        dropout = dropout,
+        batch_first = TRUE
+      )
+    }
+    
+  },
+  
+  forward = function(x) {
+    
+    x <- self$rnn(x)
+    
+    # return last states for all layers
+    # per layer, a single tensor for GRU, a list of 2 tensors for LSTM
+    x <- x[[2]]
+    x
+    
+  }
+)
+```
+
+### Attention module
+
+In basic seq2seq, whenever it had to generate a new value, the decoder took into account two things: its prior state, and the previous output generated. In an attention-enriched setup, it additionally receives the complete output from the encoder. In deciding what subset of that output should matter right now, it gets help from a new agent, the attention module.
+
+Attention may be implemented in a number of different ways. Whatever the implementation, the goal is the same: given current decoder state and well as complete encoder outputs, obtain a weighting of those outputs indicative of how relevant they are to what the decoder is currently up to. This procedure results in the so-called *attention weights*: a normalized score, for each time step in the encoding, that quantify its respective importance.
+
+Here, we show two implementation options, one additive, and one multiplicative.
+
+#### Additive variant
+
+In additive attention, encoder outputs and decoder state are commonly either added or concatenated (we choose to do the latter, below). The resulting tensor is run through a linear layer, and a softmax is applied for normalization.
+
+```{r}
+attention_module_additive <- nn_module(
+  
+  initialize = function(hidden_dim, attention_size) {
+    
+    self$attention <- nn_linear(2 * hidden_dim, attention_size)
+    
+  },
+  
+  forward = function(state, encoder_outputs) {
+    
+    # function argument shapes
+    # encoder_outputs: (bs, timesteps, hidden_dim)
+    # state: (1, bs, hidden_dim)
+    
+    # multiplex state to allow for concatenation (dimensions 1 and 2 must agree)
+    seq_len <- dim(encoder_outputs)[2]
+    # resulting shape: (bs, timesteps, hidden_dim)
+    state_rep <- state$permute(c(2, 1, 3))$repeat_interleave(seq_len, 2)
+    
+    # concatenate along feature dimension
+    concat <- torch_cat(list(state_rep, encoder_outputs), dim = 3)
+    
+    # run through linear layer with tanh
+    # resulting shape: (bs, timesteps, attention_size)
+    scores <- self$attention(concat) %>% 
+      torch_tanh()
+    
+    # sum over attention dimension and normalize
+    # resulting shape: (bs, timesteps) 
+    attention_weights <- scores %>%
+      torch_sum(dim = 3) %>%
+      nnf_softmax(dim = 2)
+    
+    # a normalized score for every source token
+    attention_weights
+  }
+)
+
+```
+
+#### Multiplicative variant
```

Comment:
```suggestion
#### Multiplicative attention
```

Reply:
ok ;-) (again, looking for variation - literally, I guess)